### PR TITLE
Clarify include guard naming convention

### PR DESCRIPTION
**Pull request by practicalswift**
* **Documentation**: Document include guard convention
* **Fix**: Fix missing or inconsistent include guards
* **Regression test**: Add `lint-include-guards.sh` which checks include guard consistency

### CODE REVIEW

**Comment by promag** [comment]
Concept ACK.

Could run script in travis where `CHECK_DOC=1` to prevent future PR's breaking it?

**Comment by practicalswift** [comment]
@promag `contrib/devtools/lint-*.sh` is executed automatically by Travis (via `lint-all.sh` which is run as part of `CHECK_DOC=1`) so `contrib/devtools/lint-include-guards.sh` already run by Travis – no need to reference it explicitly :-)

File: contrib/devtools/lint-include-guards.sh [hunk 1]
```diff
@@ -0,0 +1,24 @@
+#!/bin/bash
+
+HEADER_ID_PREFIX="BITCOIN_"
+HEADER_ID_SUFFIX="_H"
+
+REGEXP_EXCLUDE_FILES_WITH_PREFIX="src/(config/bitcoin-config.h|crypto/ctaes/|leveldb/|obj/build.h|qt/forms/|secp256k1/|tinyformat.h|univalue/)"
```
Comment (eklitzke) [inline contrib/devtools/lint-include-guards.sh]: Do you really need bitcoin-config.h and build.h in here? They're auto-generated so I would have thought git ls-files wouldn't have listed them.

File: contrib/devtools/lint-include-guards.sh [hunk 2]
```diff
@@ -0,0 +1,24 @@
+#!/bin/bash
```
Comment (eklitzke) [inline contrib/devtools/lint-include-guards.sh]: You should probably add a copyright notice to this file.

**Review by eklitzke — COMMENTED** [review]
Concept ACK

**Comment by practicalswift** [comment]
@eklitzke Thanks for reviewing! Feedback addressed. Please re-review :-)

**Comment by sipa** [comment]
utACK 3bcc0059b82b1d482d499d4d8f09118617074a5b

File: contrib/devtools/lint-include-guards.sh [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+#
+# Check include guards.
+
+HEADER_ID_PREFIX="BITCOIN_"
+HEADER_ID_SUFFIX="_H"
+
+REGEXP_EXCLUDE_FILES_WITH_PREFIX="src/(crypto/ctaes/|leveldb/|secp256k1/|tinyformat.h|univalue/)"
+
+EXIT_CODE=0
+for HEADER_FILE in $(git ls-files -- "*.h" | grep -vE "^${REGEXP_EXCLUDE_FILES_WITH_PREFIX}")
```
Comment (promag) [inline contrib/devtools/lint-include-guards.sh]: Could use `git ls-files -x` instead of `| grep -vE`?

Comment (promag) [reply contrib/devtools/lint-include-guards.sh]: Nevermind, `-x` skips untracked files.

Comment (practicalswift) [reply contrib/devtools/lint-include-guards.sh]: @promag No I don't think so (`-x` is for untracked files)? :-)

I'd be glad to be proven wrong so provide an equivalent with `git ls-files -x` and I'll switch :-)

**Comment by promag** [comment]
Tested ACK 3bcc005 on mac.

File: contrib/devtools/lint-include-guards.sh (hunk 3 above)
Comment (promag) [reply contrib/devtools/lint-include-guards.sh]: Yes I've come to that conclusion right trying it.

**Review by eklitzke — APPROVED** [review]
Thanks! utACK 3bcc0059b82b1d482d499d4d8f09118617074a5b

**Pull request merged**